### PR TITLE
executor: add test for pr 45284

### LINKED ISSUE #45607: PR 45284 needs tests

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #45607

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -860,6 +861,9 @@ func (e *IndexMergeReaderExecutor) getResultTask(ctx context.Context) (*indexMer
 
 func handleWorkerPanic(ctx context.Context, finished <-chan struct{}, ch chan<- *indexMergeTableTask, extraNotifyCh chan bool, worker string) func(r interface{}) {
 	return func(r interface{}) {
+		failpoint.Inject("testChangeProcessWorkerType", func() {
```
Comment (guo-shaoge) [inline executor/index_merge_reader.go:864]: I think this failpoint can be confusing.

Comment (guo-shaoge) [reply executor/index_merge_reader.go:864]: You can try following to reproduce this hang:
1. In `fetchLoopUnion` and `doIntersectionPerPartition`, first write `resultCh` then write `workCh`
2. After write `resultCh`, before write `workCh`, add a failpoint to call cancel func.

File: executor/index_merge_reader.go [hunk 2]
```diff
@@ -56,7 +56,8 @@ import (
 )
 
 var (
-	_ exec.Executor = &IndexMergeReaderExecutor{}
+	_                     exec.Executor = &IndexMergeReaderExecutor{}
+	TestContextCancelFunc func()
```
Comment (guo-shaoge) [inline executor/index_merge_reader.go:60]:
```suggestion
	IndexMergeCancelFuncForTest func()
```

File: executor/index_merge_reader.go (hunk 1 above)
Comment (xzhangxian1008) [reply executor/index_merge_reader.go:864]: > I think this failpoint can be confusing.

okk, I will add some comments.

File: executor/index_merge_reader.go (hunk 2 above)
Comment (xzhangxian1008) [reply executor/index_merge_reader.go:60]: > 

okk

**Review by guo-shaoge — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**